### PR TITLE
Hot Exit gets hotter

### PR DESCRIPTION
**Pull request by bpasero**
* Investigate to restore untitled dirty files via editor input factories (https://github.com/Microsoft/vscode/issues/15498)
* Hot Exit: Untitled files with no content should be restored (https://github.com/Microsoft/vscode/issues/14690)
* Untitled files are not restored to their previous groups (https://github.com/Microsoft/vscode/issues/14116)
* Fix dirty tab indicator after hot exit restore (https://github.com/Microsoft/vscode/issues/14038)

### CODE REVIEW

**Comment by bpasero** [comment]
@Tyriar ready for some feedback from you. lots of changes included:
* untitled restore now via the editor input factory
* untitled backup restore moved into untitled editor model
* `backupRestorer` is the workbench contribution that handles the restore after startup
* some fixes so that the ugly timeout is not needed anymore for showing dirty

The biggest hack in this change is the fact that I need to walk over all opened editors to find the ones with backup when instead I would like to have a method on `IBackupFileService` to give me a list of files and untitled with backups. As soon as we have such a method we can use the backup restorer even when hot exit is disabled and it would open the files with backups after a crash independent from what the user has opened (because that is likely stale after a crash).

In the state as it is now I nicely see proper dirty indication for files and untitled and things open in the right location where I left off.

File: src/vs/workbench/parts/backup/common/backupRestorer.ts [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+import { IBackupService, IBackupFileService } from 'vs/workbench/services/backup/common/backup';
+import { IEditorGroupService } from 'vs/workbench/services/group/common/groupService';
+import { FileEditorInput } from 'vs/workbench/parts/files/common/files';
+
+// TODO@ben TODO@tyriar this should restore any backup that exists on disk and not rely
+// on the editors to be restored already in the stacks model. For that a method is needed
+// to get all backups that exist on disk.
+export class BackupRestorer implements IWorkbenchContribution {
+
+	public _serviceBrand: any;
```
Comment (Tyriar) [inline src/vs/workbench/parts/backup/common/backupRestorer.ts]: Not needed

**Review by Tyriar — APPROVED** [review]
Looks good to me!

File: src/vs/workbench/parts/backup/common/backupRestorer.ts (hunk 1 above)
Comment (bpasero) [inline src/vs/workbench/parts/backup/common/backupRestorer.ts]: Thanks

**Pull request merged**